### PR TITLE
Add GC call after evaluating an objective function.

### PR DESCRIPTION
**Pull request by toshihikoyanase**
This PR adds an invocation of garbage collection after evaluating an objective function. This fix reduces memory footprints when the programs are executed in VM environments like CircleCI as reported in #325.

---

## Memory Usage Comparison

The memory usage of the chainermn example is about 0.9GB as reported in #325. The memory usage of this PR is about 0.7GB, and it seems to work as expected.

```bash
// Preparation.
$ ssh -p ${CIRCLE_CI_PORT} ${CIRCLE_CI_HOST}
$ cd project && . venv/bin/activate
$ STORAGE_URL=sqlite:///example.db
$ STUDY_NAME=`optuna create-study --storage $STORAGE_URL`
$ export OMP_NUM_THREADS=1
$ sudo apt install -y time

// Execute a ChainerMN example.
$ /usr/bin/time -v mpirun -n 2 -- python examples/chainermn_simple.py $STUDY_NAME $STORAGE_URL | grep 'Maximum resident'
	Maximum resident set size (kbytes): 759408  // Memory Usage=0.7GB
	Maximum resident set size (kbytes): 749364  // Memory Usage=0.7GB
	Maximum resident set size (kbytes): 751100  // Memory Usage=0.7GB
```

### CODE REVIEW

File: optuna/integration/chainermn.py [hunk 1]
```diff
@@ -147,6 +148,11 @@ def optimize(
                     pass
                 except catch:
                     pass
+                finally:
+                    # The following line mitigates memory problems due to container virtualization.
```
Comment (sile) [inline optuna/integration/chainermn.py]: I think that the problem being addressed in #325 is not specific to container virtualization environments (at least in theory). Therefore, I prefer an expression that covers a broader range such as "memory problems that can be occurred in some environments".

File: optuna/integration/chainermn.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 from __future__ import absolute_import
 
+import gc
```
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/integration/chainermn.py]: You are right. I'll add it.

File: optuna/integration/chainermn.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/integration/chainermn.py]: Thank you for your suggestion. I mostly agree with you, but I feel that 'some environments' is ambiguous.
I think adding an example like 'some environments (e.g., services that use computing containers such as CircleCI)'  is useful for developers.
What do you think?

Comment (sile) [reply optuna/integration/chainermn.py]: Sounds good.

Comment (toshihikoyanase) [reply optuna/integration/chainermn.py]: Thank you. I'll fix the comments.

File: optuna/integration/chainermn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/integration/chainermn.py]: I've fixed this issue at the commit https://github.com/pfnet/optuna/pull/377/commits/3fd988a8ef8b4c97dd58ce56fba305f1c98f6ef1.

File: optuna/integration/chainermn.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/integration/chainermn.py]: I've updated the comments at the commit https://github.com/pfnet/optuna/pull/377/commits/0b8e509c5f24d5ad2120a5592af8940126ba9a9c.

**Pull request merged**